### PR TITLE
Add DAE set manipulation utilities

### PR DESCRIPTION
**Pull request by Robbybp**
## Summary/Motivation:
Methods for identifying when a component is (explicitly, implicitly) indexed by a set, and for identifying the 'projection' of a SetProduct into a 'lower-dimension' set (e.g. onto a set with a specified 'time' value) are useful for manipulating DAE models, i.e. in integration and NMPC.

## Changes proposed in this PR:
-Function for identifying when a component is explicitly indexed by a set
-Function for identifying when a component is implicitly indexed by a set
-Functions for identifying index sets 'except' certain sets
-Tests

### Legal Acknowledgement

By contributing to this software project, I agree to the following terms and conditions for my contribution:

1. I agree my contributions are submitted under the BSD license.
2. I represent I am authorized to make the contributions and grant the license. If my employer has rights to intellectual property that includes these contributions, I represent that I have received permission to make contributions and grant the required license on behalf of that employer.

### CODE REVIEW

**Comment by qtothec** [comment]
Only comment is whether `misc` is the best name for indicating modeling utilities.

**Comment by Robbybp** [comment]
Just added to new file set_utils. Maybe some related pre-existing functions from misc like get_index_information should go in set_utils as well.

File: pyomo/dae/set_utils.py [hunk 1]
```diff
@@ -0,0 +1,173 @@
...
+        # Continue up the tree, checking the parent block of our potentially-
+        # indexed block, which I would like to assume contains the BlockData
+        # we started from.
+        else:
+            parent = parent.parent_block()
+    # Return False if top-level block was reached.
+    return False
+
+
+def get_index_set_except(comp, *sets):
```
Comment (jsiirola) [inline pyomo/dae/set_utils.py]: This function seems to duplicate what can be done with a `Reference`.  Can you elaborate more on the expected usage here?  In particular, the use of the `index_getter` function?

File: pyomo/dae/tests/test_set_utils.py [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+
+        # Want this to give a TypeError
+        # info = get_index_set_except(m.v, m.time)
+
+        # Indexed by one set
+        info = get_index_set_except(m.v1, m.time)
+        set_except = info['set_except']
+        index_getter = info['index_getter']
+        self.assertTrue(set_except == [None])
+        self.assertEqual(index_getter([5,6,7], 3j), 3j)
```
Comment (jsiirola) [inline pyomo/dae/tests/test_set_utils.py]: Why are you using complex numbers here?

File: pyomo/dae/tests/test_set_utils.py [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+
+        # Indexed by two sets
+        info = get_index_set_except(m.v2, m.time)
+        set_except = info['set_except']
+        index_getter = info['index_getter']
+        self.assertTrue(m.space[1] in set_except
+                        and m.space.last() in set_except)
+        self.assertEqual(index_getter(2, 4), (4, 2), index_getter((2,), 4))
```
Comment (jsiirola) [inline pyomo/dae/tests/test_set_utils.py]: What are you testing here?  it looks like you are providing `index_getter((2,),4)` the failure message?  If that is your intent, please at lease form a string message that would help a developer to understand what went wrong.

Comment (jsiirola) [reply pyomo/dae/tests/test_set_utils.py]: OK - your comment helps.  However, that should be two separate `assertEqual` statements (one for `(2,4)` and one for `((2,),4)`).  Right now, you are providing `index_getter((2,), 4)` as the `message` argument to `assertEqual`.

File: pyomo/dae/set_utils.py (hunk 1 above)
Comment (Robbybp) [reply pyomo/dae/set_utils.py]: This function is heavily inspired by `get_index_information` in `dae.misc`. Its intended use is for components with the same number of indexing sets, some of which are identical. The non-identical sets are intended but not required to be continuous, differing by bounds, discretization, etc. 

The goal is, given some elements for the non-identical sets, to find every index of the components for which these non-identical sets (time and space usually) take on the values specified. The usage is:

```python
if is_explicitly_indexed_by(var1, time):
    info = get_index_set_except(var1, time)
    non_time_set = info['set_except']
    index_getter = info['index_getter']
    for partial_index in non_time_set:
        new_index = index_getter(partial_index, new_time_value)
        # do something with new_index
```

`index_getter` inserts values for the sets that have been excluded into partial indices to form a new index with the new values for these sets.

This is useful for
- initializing dynamic models to values from a corresponding steady state model
- initializing dynamic models to their initial conditions 
- copying values between two dynamic (or PDE) models of different discretizations

You could probably do this with references, something like:
```python
for index in Reference(var1[t0, ...]): # requires knowledge of where time is in var's index
    # index is a partial index excluding time
    # need to insert new time value into index
    var2[new_index].set_value(var1[old_index])
```
 I think I remember trying, running into a snag somewhere, and going with this approach instead as it was more similar to what I already had implemented.

File: pyomo/dae/set_utils.py [hunk 4]
```diff
@@ -0,0 +1,173 @@
...
+
+    if comp.dim() == 1:
+        # In this case, assume that comp is indexed by *sets
+        # Return the trivial set_except and index_getter
+        info['set_except'] = [None]
+        # index_getter here will only accept a single argument
+        info['index_getter'] = (lambda incomplete_index, newval: newval)
+        return info
+
+    set_tuple = comp.index_set().set_tuple
```
[the remark this answers is not in the document]
Comment (Robbybp) [reply pyomo/dae/set_utils.py]: I hadn't considered this. Is there a better way check for indexing by a `SetProduct`? I may need to limit this function to only allow "base" sets of dimension 1

[the remark this answers is not in the document]
Comment (jsiirola) [reply pyomo/dae/set_utils.py]: Right now it is challenging.  If I can finish up the set merge, set_tuple is being replaced by a `subsets()` method that will be available on *all* set objects, so you would just use that.

[the remark this answers is not in the document]
Comment (Robbybp) [reply pyomo/dae/set_utils.py]: In that case maybe this PR should be postponed until then. I can add these methods, with a check like `hasattr(comp.index_set(), 'set_tuple')`, to `idaes.core.util.dyn_utils` so the integrator can move forward, and I can rewrite/reconsider this when the set merge happens.

[the remark this answers is not in the document]
Comment (jsiirola) [reply pyomo/dae/set_utils.py]: I think that makes sense.  You can either check for the existence of set_tuple (probably the best), or if the set is an instance of SetProduct.

File: pyomo/dae/tests/test_set_utils.py [hunk 5]
```diff
@@ -0,0 +1,207 @@
...
+from pyomo.environ import *
+from pyomo.common.log import LoggingIntercept
+from pyomo.dae import *
+from pyomo.dae.set_utils import *
+from pyomo.core.kernel.component_map import ComponentMap
+
+currdir = dirname(abspath(__file__)) + os.sep
+
+
+class TestDaeSetUtils(unittest.TestCase):
```
Comment (blnicho) [inline pyomo/dae/tests/test_set_utils.py:30]: Could you add a few more tests for components that are also indexed by multi-dimensional Sets and components where the ContinuousSet is not always the first indexing set? Looks like you have these kinds of tests for one function but not the other. I've seen edge cases and bugs crop up with multi-dimensional sets and set ordering when implementing similar functionality.

File: pyomo/dae/set_utils.py [hunk 6]
```diff
@@ -0,0 +1,217 @@
...
+    """
+    if not comp.is_indexed():
+        return False
+    for s in sets:
+        if isinstance(s, SetProduct):
+            msg = ('Checking for explicit indexing by a SetProduct '
+                  'is not supported')
+            raise TypeError(msg)
+
+    expand_all_set_operators = kwargs.pop('expand_all_set_operators', False)
```
[the remark this answers is not in the document]
Comment (Robbybp) [reply pyomo/dae/set_utils.py]: It is now an optional argument.

File: pyomo/dae/tests/test_set_utils.py (hunk 5 above)
Comment (Robbybp) [reply pyomo/dae/tests/test_set_utils.py:30]: I believe I have implemented satisfactory tests.

File: pyomo/dae/set_utils.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (blnicho) [reply pyomo/dae/set_utils.py]: Oops, I didn't notice that your original implementation was using both `*args` and `**kwargs`. In that case you actually do need to stick with `**kwargs` because I don't think Python 2.7 supports `*args` followed by named keyword arguments without using `**kwargs` (as you'll see in the failing tests). Could you revert this change except for the docstring?

[the remark this answers is not in the document]
Comment (Robbybp) [reply pyomo/dae/set_utils.py]: Yeah I see that in the failing test, but when I tried locally, it seemed to work with 2.7. I'll revert it.

**Comment by blnicho** [comment]
@jsiirola could you take another look at this one? I think it's ready but we need one more approving review.
